sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,372 @@
+    }
+    outputTimersReceivers.put(timerKey, receiver);
+    return receiver;
+  }
+
+  public void flush() throws IOException {
+    Elements.Builder elements = convertBufferForTransmission();
+    if (elements.getDataCount() > 0 || elements.getTimersCount() > 0) {
+      outboundObserver.onNext(elements.build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  public void sendBufferedDataAndFinishOutboundStreams() {
+    if (totalByteCounter.longValue() == 0
+        && outputTimersReceivers.isEmpty()
+        && outputDataReceivers.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    LOG.debug(
+        "Closing streams for outbound endpoints {} and {}.",
+        outputDataReceivers.keySet(),
+        outputTimersReceivers.keySet());
+    LOG.debug(
+        "Sent outbound data : {}, outbound timers : {}.",
+        outputDataReceivers,
+        outputTimersReceivers);
+    for (String pTransformID : outputDataReceivers.keySet()) {
+      bufferedElements
+          .addDataBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(pTransformID)
+          .setIsLast(true);
+    }
+    for (TimerEndpoint timerKey : outputTimersReceivers.keySet()) {
+      bufferedElements
+          .addTimersBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(timerKey.pTransformId)
+          .setTimerFamilyId(timerKey.timerFamilyId)
+          .setIsLast(true);
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    outputDataReceivers.values().forEach(Receiver::reset);
+    outputTimersReceivers.values().forEach(Receiver::reset);
+  }
+
+  private Elements.Builder convertBufferForTransmission() {
+    Elements.Builder bufferedElements = Elements.newBuilder();
+    for (Map.Entry<String, Receiver<?>> entry : outputDataReceivers.entrySet()) {
+      if (entry.getValue().getOutput().size() == 0) {
+        continue;
+      }
+      ByteString bytes = entry.getValue().getOutput().toByteString();
+      bufferedElements
+          .addDataBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(entry.getKey())
+          .setData(bytes);
+      entry.getValue().reset();
+    }
+    for (Map.Entry<TimerEndpoint, Receiver<?>> entry : outputTimersReceivers.entrySet()) {
+      if (entry.getValue().getOutput().size() == 0) {
+        continue;
+      }
+      ByteString bytes = entry.getValue().getOutput().toByteString();
+      bufferedElements
+          .addTimersBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(entry.getKey().pTransformId)
+          .setTimerFamilyId(entry.getKey().timerFamilyId)
+          .setTimers(bytes);
+      entry.getValue().reset();
+    }
+    totalByteCounter.reset();
```

Comment:
Since `convertBufferForTransmission` should always be accessed from either a single thread or synchronized via the flushLock we can turn `totalByteCounter` into a normal long.